File: Nemeio.WinAutoInstaller/ErrorMessageProvider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Nemeio.WinAutoInstaller.Models;

namespace Nemeio.WinAutoInstaller
{
    public class ErrorMessageProvider
    {
        private Dictionary<ErrorCode, string> _errorMessage;

        public ErrorMessageProvider()
        {
            _errorMessage = new Dictionary<ErrorCode, string>()
            {
                { ErrorCode.WinAutoInstallerKeyboardNotFound, "No keyboard found." },
                { ErrorCode.WinAutoInstallerInternetNotAvailable, "An Internet connection is required to download Nemeio. Please try again later." },
                { ErrorCode.WinAutoInstallerServerTimeout, "No response from the server within the allotted time. Please try again later." },
                { ErrorCode.WinAutoInstallerDownloadFailed, "Download failed." },
                { ErrorCode.WinAutoInstallerDownloadCancelByUser, "You have stopped the installer." },
                { ErrorCode.WinAutoInstallerInvalidChecksum, "Downloaded file is corrupted. Please try again." },
                { ErrorCode.WinAutoInstallerGetDownloadInfoFailed, "Can't request installer server. Please try again later." },
                { ErrorCode.WinAutoInstallerFileSystemWriteDenied, "Unauthorized to write on file system." },
                { ErrorCode.WinAutoInstallerUpdateNotFound, "No installer found. Please contact support." },
            };
        }

        public string GetErrorMessage(ErrorCode code)
        {
            if (_errorMessage.ContainsKey(code))
            {
                return _errorMessage[code];
            }

            return "An error occured. Please try again later.";
        }

        public string GetFullErrorMessage(ErrorCode code)
        {
            var errorCode = (int)code;
            var errorMessage = GetErrorMessage(code);

            var hexadecimalErrorCode = errorCode.ToString("X02");

            return $"[{hexadecimalErrorCode}] {errorMessage}";
        }
    }
}

File: Nemeio.MacAutoInstaller/Nemeio.MacAutoInstaller/Models/ErrorMessageProvider/ErrorMessageProvider.m
//
//  ErrorMessageProvider.m
//  Nemeio.MacAutoInstaller
//
//  Created by Kevin Sibue on 11/02/2020.
//  Copyright © 2020 Kevin Sibue. All rights reserved.
//

#import "ErrorMessageProvider.h"

@implementation ErrorMessageProvider

const NSString* DEFAULT_ERROR_MESSAGE = @"An error occured. Please try again later.";

- (instancetype)init {
    _errorMessages = [NSMutableDictionary dictionaryWithObjectsAndKeys:
      @"No keyboard found.", [NSNumber numberWithInt:MAC_AUTO_INSTALLER_KEYBOARD_NOT_FOUND],
      @"An Internet connection is required to download Nemeio. Please try again later.", [NSNumber numberWithInt:MAC_AUTO_INSTALLER_INTERNET_NOT_AVAILABLE],
      @"No response from the server within the allotted time. Please try again later.", [NSNumber numberWithInt:MAC_AUTO_INSTALLER_SERVER_TIMEOUT],
      @"Download failed.", [NSNumber numberWithInt:MAC_AUTO_INSTALLER_DOWNLOAD_FAILED],
      @"You have stopped the installer.", [NSNumber numberWithInt:MAC_AUTO_INSTALLER_DOWNLOAD_CANCEL_BY_USER],
      @"Downloaded file is corrupted. Please try again.", [NSNumber numberWithInt:MAC_AUTO_INSTALLER_INVALID_CHECKSUM],
      @"Can't request installer server. Please try again later.", [NSNumber numberWithInt:MAC_AUTO_INSTALLER_GET_DOWNLOAD_INFO_FAILED],
      @"Unauthorized to write on file system.", [NSNumber numberWithInt:MAC_AUTO_INSTALLER_FILE_SYSTEM_WRITE_DENIED],
      @"No installer found. Please contact support.", [NSNumber numberWithInt:MAC_AUTO_INSTALLER_UPDATE_NOT_FOUND],
    nil];
    return self;
}

- (NSString*)getErrorMessage: (ErrorCode)code {
    NSNumber *codeNumber = [NSNumber numberWithInt: code];
    NSString *message = [_errorMessages objectForKey: codeNumber];
    if (!message) {
        return DEFAULT_ERROR_MESSAGE;
    }
    return message;
}

@end
